genesis_core/user_api/iam/api/controllers.py

```
@@ -44,6 +47,33 @@ def enforce(self, rule, do_raise=False, exc=None):
         return iam.enforcer.enforce(rule, do_raise, exc)
 
 
+class ValidateMixin:
+    vm_min_length = 8
+    vm_not_contain: str = string.whitespace
+    vm_must_contain: list[str] = None  # [digits, ascii_uppercase, punctuation]
+    vm_regex: str = None
+
+    def validate(self, text):
+        if text is None:
+            return
+        if (
+            (self.vm_min_length and len(text) < self.vm_min_length)
+            or (self.vm_not_contain and set(self.vm_not_contain) & set(text))
+            or (
+                self.vm_must_contain
+                and not all(
+                    [set(part) & set(text) for part in self.vm_must_contain]
+                )
+            )
+            or (self.vm_regex and re.match(self.vm_regex, text))
```

The regex check might be incorrect. `re.match()` returns a match object (truthy) on success and `None` (falsy) on failure, so the condition should be `not re.match()` when you want to raise an error on mismatch. Also, might be worth adding an explicit comment near `vm_regex: str = None` saying smth like "string MUST match this regex".

---

fixed and added error text